docs/build/smart-contracts/example-contracts/bls-signature.mdx

```diff
@@ -0,0 +1,531 @@
+
+        // Perform the pairing check, i.e. e(pk, msg)*e(-g1, sig) == 1, which is
+        // equivalent to checking `e(pk, msg) == e(g1, sig)`.
+        // The LHS = e(sk * g1, msg) = sk * e(g1, msg) = e(g1, sk * msg) = e(g1, sig),
+        // thus it must equal to the RHS if the signature matches.
+        if !bls.pairing_check(vp1, vp2) {
+            return Err(AccError::InvalidSignature);
+        }
+        Ok(())
+    }
+}
+```
+
+The `CustomAccountInterface::__check_auth` function implements the custom signature verification logic for this account.
+
+`env.crypto().bls12_381()` initializes the bls12_381 module from which the BLS12-381 functions are available. The `signature_payload` contains the payload that was signed.
+
+`bls.hash_to_g2(&signature_payload.into(), &dst)` hashes the message into the G2 group. `agg_sig` contains the aggregate signature which is another point in G2.
+
+To perform the signature verification, we construct two vectors `Vec<G1Affine>` and `Vec<G2Affine>`, and call `bls.pairing_check` on them. The pairing check function performs `e(pk, msg)*e(-g1, sig) == 1` which is equivalent to checking `e(pk, msg) == e(g1, sig)`.
+
+### Tests
+
+Open the [`bls_signature/src/test.rs`] file to follow along.
+
+[`auth/src/test.rs`]: https://github.com/stellar/soroban-examples/tree/v22.0.1/bls_signature/src/test.rs
+
+```rust title="auth/src/test.rs"
+#[test]
+fn test() {
+    let env = Env::default();
+    env.budget().reset_unlimited();
+    let pk = aggregate_pk_bytes(&env);
+    env.mock_all_auths();
+
+    let client = create_client(&env);
+    client.init(&pk);
+    let payload = BytesN::random(&env);
+    let sig_val = sign_and_aggregate(&env, &payload.clone().into()).to_val();
+
+    env.budget().reset_default();
+    env.try_invoke_contract_check_auth::<AccError>(&client.address, &payload, sig_val, &vec![&env])
+        .unwrap();
+    env.budget().print();
+}
+```
+
+Most of what's here is needed in order to create the contract client and ensure the invocation of its custom account interface for signature authorization. After the setup, calling `env.try_invoke_contract_check_auth` on the client will invoke the `__check_auth` logic we've defined in our contract.
+
+The invocation will return nothing on success, and will panic on failure. The other thing we are interested here is budget consumption for this call. At the beginning setting `env.budget().reset_unlimited()` ensure all the prep work (which emulates the off-chain work needed to be performed by e.g. a smart wallet) will not run out of budget. This isn't strictly necessary since the default budget should be large enough. `env.budget().reset_default()` right before the invocation makes sure the budget is reset to default state (also resetting metering counters) so that we can get an accurate capture of the budget for the invocation alone. `env.budget().print()` at the end prints out the budget.
```

Comment:
Since SDK 22.0.4 we shouldn't need to do budget resets anymore. We should update the example first, but I think for the doc we can just omit this, so that there is less to change when the example is updated.